### PR TITLE
Cpu profiles (1/3): CPUID adjustments

### PR DESCRIPTION
**Pull request by olivereanderson**
This is the first of three PRs that have been split off from https://github.com/cyberus-technology/cloud-hypervisor/pull/25

This PR contains all the functionality needed to launch CHV with a CPU profile (currently limited to Intel Skylake and Sapphire rapids processors), minus the fact that MSR restrictions are still missing.

The next two PRs will contain CPUID definitions and the CPU profiles generation tool respectively.

Note: This contains a lot of code, because of the generated JSON files corresponding to the CPU profiles. The JSON files are currently pretty printed in order to make debugging easier at this relatively early stage.

# Steps to Undraft
- [ ] Wait for patchset v50 + rebase

### Why do we even need this?

Recall that software is usually developed to run on a variety of processors  with various features. In order for the software to dynamically discover which hardware features may be utilized one typically uses the CPUID instruction to query the CPU for information. In some cases one can also obtain CPU information via so called MSRs (model specific registers), but we will leave those out of this discussion for the time being.

Consider now the case that a guest is running some workload on host A which then gets migrated to host B where host B has a different processor than A. If this is a live migration (i.e. it is performed while the software is running), then the guest's workload can easily run into a time of check to time of use error. 

Luckily hypervisors are able to manipulate what CPUID returns to guests when called which we can and will take advantage of in order to make live migration safer.  Indeed if there is a subset of CPU features and properties that
are shared by all CPUs in a cluster, then if all hosts restrict themselves to that subset then live migration suddenly becomes a lot less problematic (although still not an entirely solved problem). 

#### Using an existing CPU profile
In this PR we patch cloud hypervisor so that users may specify a CPU profile on the command line. This will in turn restrict what CPUID values the guest obtains and may thus affect functionality and performance, but may still be the best tool in the box when live migration to (subtly) different hardware is desired.

From a user perspective one simply includes the `profile=<desired cpu profile>` argument to the `cpus` parameter e.g.
```
--cpus boot=8,profile=sapphire-rapids
```
in order to utilize the Intel Sapphire rapids server CPU profile. If the host has hardware considered to be compatible with the chosen profile, then cloud hypervisor should work just like before, with the exception that guest's may see certain other values when inspecting CPUID.

### The design in a bit of detail

The implementation is based on the understanding that we do not only need to work with bit sets, but we also need to manipulate some multi-bit values. Indeed, some CPUID output values indicate how many leaves there are (such as for instance leaf 0, sub-leaf 0, EAX), while others may tell you the number of sub-leaves, or the size of a state component for instance. Note that if the CPUID instruction is executed with an invalid leaf, some processors will return the data for the highest basic information leaf. In a live-migration setting this could easily lead to a time of check to time of use error!

Since we cannot simply deal with bit sets (at least to the best of our understanding) we unfortunately end up with a rather more complex solution than what we had initially hoped for. 

#### CPU Profile policies

The idea is that we have a static list describing all known values one can obtain from CPUID (on an Intel CPU) and also such a list of the values defined by KVM (within the leaf range reserved for hypervisors).  Note that this list is **not included in this PR**, but will be the part of our next PR.

From the aforementioned list of CPUID values we built a tool to generate CPU profiles when executed on the hardware that one wishes to be compatible with. The CPU profile generation tool (which will be revealed in a follow up PR) produces
data (currently JSON files) that we load at runtime and use to adjust the CPUID entries that guests will see when executing the CPUID instruction.

### Immediate Follow up tasks

- [ ] Introduce a least common denominator test that can run in CI.
- [ ] Make CPU profiles take MSR restrictions into account. The first approximation would perhaps be to store whatever is returned when the compatibility_target calls [KVM_GET_MSR_FEATURE_INDEX_LIST]https://docs.kernel.org/virt/kvm/api.html#kvm-get-msr-index-list-kvm-get-msr-feature-index-list) and set a [filter](https://docs.kernel.org/virt/kvm/api.html#kvm-x86-set-msr-filter) that by default denies anything not listed there.
- [ ] Generate more thorough migration/cpuid compatibility checks than what (already exists upstream) by using the `MigrationCompatibilityRequirements` in the cpuid definitions introduced here.

### CODE REVIEW

File: hypervisor/src/arch/x86/mod.rs [hunk 1]
```diff
@@ -396,9 +396,7 @@ impl XsaveState {
                 ((size as usize) - size_of::<kvm_bindings::kvm_xsave>())
                     .div_ceil(size_of::<kvm_bindings::__u32>())
             };
-            XSAVE_FAM_LENGTH
-                .set(fam_length)
-                .expect("This should only be set once");
+            let _ = XSAVE_FAM_LENGTH.set(fam_length);
```
Comment (phip1611) [inline hypervisor/src/arch/x86/mod.rs:399]: commit message `Temporary workaround until we switch over to the WIP fix upstream` - is this intentional?

File: arch/src/x86_64/cpuid_definitions/mod.rs [hunk 2]
```diff
@@ -0,0 +1,40 @@
+use std::io::Write;
+use std::ops::RangeInclusive;
+
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+use crate::x86_64::CpuidReg;
+
+pub(in crate::x86_64) fn serialize_as_hex<S: Serializer>(
```
Comment (phip1611) [inline arch/src/x86_64/cpuid_definitions/mod.rs:13]: I think it would be great to add unit tests for this function

File: arch/src/x86_64/cpuid_definitions/mod.rs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+) -> Result<S::Ok, S::Error> {
+    // two bytes for "0x" prefix and eight for the hex encoded number
+    let mut buffer = [0_u8; 10];
+    let _ = write!(&mut buffer[..], "{:#010x}", input);
+    let str = core::str::from_utf8(&buffer[..])
+        .expect("the buffer should be filled with valid UTF-8 bytes");
+    serializer.serialize_str(str)
+}
+
+pub(in crate::x86_64) fn deserialize_from_hex<'de, D: Deserializer<'de>>(
```
Comment (phip1611) [inline arch/src/x86_64/cpuid_definitions/mod.rs:25]: same here, please add unit tests covering happy paths and corner cases

File: hypervisor/src/arch/x86/mod.rs (hunk 1 above)
Comment (olivereanderson) [reply hypervisor/src/arch/x86/mod.rs:399]: Yes.

Once https://github.com/cloud-hypervisor/cloud-hypervisor/pull/7534 is merged then we can remove this code.

File: arch/src/x86_64/cpu_profile.rs [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+use hypervisor::{CpuVendor, HypervisorType};
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use crate::x86_64::CpuidReg;
+use crate::x86_64::cpuid_definitions::{Parameters, deserialize_from_hex, serialize_as_hex};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
+#[serde(rename_all = "kebab-case")]
+#[allow(non_camel_case_types)]
```
Comment (phip1611) [inline arch/src/x86_64/cpu_profile.rs]: I don't see a non camel case type here?

File: arch/src/x86_64/cpu_profile.rs [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+    pub(in crate::x86_64) fn data(&self, amx: bool) -> Option<CpuProfileData> {
+        let mut data: CpuProfileData = match self {
+            Self::Host => None,
+            Self::Skylake => todo!(),
+            Self::SapphireRapids => todo!(),
+        }?;
+
+        if !amx {
+            // In this case we will need to wipe out the AMX tile state components (if they are included in the profile)
+            for adj in data.adjustments.iter_mut() {
```
Comment (phip1611) [inline arch/src/x86_64/cpu_profile.rs:58]: would it make sense to have some sort of a generic `wipe out bits from CPU id` function here?

File: arch/src/x86_64/mod.rs [hunk 6]
```diff
@@ -128,6 +130,10 @@ pub enum Error {
     #[error("Error getting supported CPUID through the hypervisor API")]
     CpuidGetSupported(#[source] HypervisorError),
 
+    /// Error when attempting to apply a CPU profile during common CPUID generation.
+    #[error("The desired CPU profile cannot be utilized due to an incompatibility issue")]
+    CpuProfileIncompatibility(#[source] anyhow::Error),
```
Comment (phip1611) [inline arch/src/x86_64/mod.rs]: Please don't! We decided to not accept any further `anyhow in thiserror` error modelling in Cloud Hypervisor as it combines the worst of both worlds. 

Please introduce a new error enum variant.

File: arch/src/x86_64/mod.rs [hunk 7]
```diff
@@ -615,167 +626,226 @@ pub fn generate_common_cpuid(
...
+        host_adjusted_to_profile.map_err(|e| Error::CpuProfileIncompatibility(e.into()))?;
+
+    // There should be relatively few cases where live migration can succeed between hosts from different
+    // CPU vendors and making our checks account for that possibility would complicate things substantially.
+    // We thus require that the host's cpu vendor matches the one used to generate the CPU profile.
+    if let Some(profile_cpu_vendor) = profile_cpu_vendor
+        && profile_cpu_vendor != hypervisor.get_cpu_vendor()
+    {
+        return Err(Error::CpuProfileIncompatibility(anyhow!(
+            "Unable to utilize CPU profile: CPU vendor mismatch detected"
```
Comment (phip1611) [inline arch/src/x86_64/mod.rs]: As above. Please introduce a new error enum variant. No anyhow.

File: arch/src/x86_64/mod.rs [hunk 8]
```diff
@@ -615,167 +626,226 @@ pub fn generate_common_cpuid(
...
+    if !use_custom_profile {
+        Ok(host_cpuid)
+    } else {
+        // Final compatibility checks to ensure that the CPUID values we return are compatible both with the CPU profile and the host we are currently running on.
+        let host_adjusted_to_profile = host_adjusted_to_profile.expect("The profile adjusted cpuid entries should exist as we checked that we have a custom CPU profile");
 
-    Ok(cpuid)
+        // Check that the host's cpuid is indeed compatible with the adjusted profile. This is not by construction.
+        info!("checking compatibility between host adjusted to profile and the host itself");
+        CpuidFeatureEntry::check_cpuid_compatibility(&host_adjusted_to_profile, &host_cpuid).context("Unable to adjust the host to the CPU profile. The resulting cpuid is not compatible with the host's cpuid entries").map_err(Error::CpuProfileIncompatibility)?;
```
Comment (phip1611) [inline arch/src/x86_64/mod.rs]: I think you should run `cargo +nighty fmt --all` on every commit. This would be reformatted I think.

**Review by phip1611 — CHANGES_REQUESTED** [review]
Amazing, outstanding work! Great git commit history. I left a few remarks.

File: arch/src/x86_64/cpu_profile.rs (hunk 5 above)
Comment (olivereanderson) [reply arch/src/x86_64/cpu_profile.rs:58]: I would prefer not to make drastic changes like that at this stage. This is all rather complex and there is no clean way of doing things, at least not without refactoring the world.

One relatively simple alternative would be to have two profiles for every supported CPU (one with and one without AMX). Then this wouldn't be necessary at all, but if more complicated features like that get introduced in the future you end up with a combinatorial explosion of profiles that need to be generated.

Comment (phip1611) [reply arch/src/x86_64/cpu_profile.rs:58]: > at least not without refactoring the world.

👍🏻  alright, i'm fine with it

**Review by phip1611 — APPROVED** [review]
LGTM!

File: arch/src/x86_64/cpu_profile.rs [hunk 9]
```diff
@@ -18,7 +18,9 @@ use crate::x86_64::cpuid_definitions::{Parameters, deserialize_from_hex, seriali
 pub enum CpuProfile {
     #[default]
     Host,
+    #[cfg(feature = "kvm")]
```
Comment (phip1611) [inline arch/src/x86_64/cpu_profile.rs:21]: is it the right thing to feature gate them for KVM?

Comment (olivereanderson) [reply arch/src/x86_64/cpu_profile.rs:21]: Yes and no.

Adjusting CPUID to what is set in the profile is in principle hypervisor agnostic. However, the adjustments error if the hypervisor's call to `Hypervisor::get_supported_cpuid` doesn't return entries for leaf values the CPU profile expects.

If there is a big drift between which CPU features are supported by KVM and MSHV, then since we are currently only generating CPU profiles with KVM as an input to the generation tool, it would then be unlikely that the CPU profiles work with the MSHV hypervisor and it is perhaps then better not to advertise them to the user in that case.

I see two ways forward:
1. Generate CPU profiles with both hypervisors. We can definitely make this invisible to the user API and just select the profile corresponding to the hypervisor that is currently in use at runtime. If we choose this path then I think it is also OK that some profiles may only be available for one hypervisor.
2. We hope that the two hypervisors support essentially the same set of CPU features (at least among those that make sense in terms of live migration) and we just keep generating CPU profiles with whatever KVM supports.

Comment (phip1611) [reply arch/src/x86_64/cpu_profile.rs:21]: Please reach out to Thomas and ask him about this. We do not have to take action in this PR but I think Thomas knows more background here and we can then persist this in some form of ticket

**Comment by olivereanderson** [comment]
Closing in favor of https://github.com/cyberus-technology/cloud-hypervisor/pull/61
